### PR TITLE
Make exception handling more consistent

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Please see the commit messages for details.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/868)

### CODE REVIEW

File: utils/src/main/kotlin/Utils.kt [hunk 1]
```diff
@@ -392,6 +392,18 @@ fun Throwable.collectMessages(): List<String> {
     return messages
 }
 
+/**
+ * Recursively collect the messages of this [Throwable] and all its causes and join them to a single [String].
+ */
+fun Throwable.collectMessagesAsString() =
```
Comment (sschuberth) [inline utils/src/main/kotlin/Utils.kt]: Couldn't we simply use `Throwable.printStackTrace(PrintWriter s)`? Also see https://stackoverflow.com/a/4812589/1127485.

Comment (sschuberth) [reply utils/src/main/kotlin/Utils.kt]: Ah, right!

File: utils/src/main/kotlin/Utils.kt [hunk 2]
```diff
@@ -392,6 +392,18 @@ fun Throwable.collectMessages(): List<String> {
...
+        collectMessages().let { messages ->
+            if (messages.size == 1) {
+                messages.first()
+            } else {
+                messages.joinToString(", caused by ") { "\"$it\"" }
```
Comment (sschuberth) [inline utils/src/main/kotlin/Utils.kt]: Another nit: Should we still align to how `printStackTrace()` writes "Caused by:" in a new line, see r.g. https://ci.appveyor.com/project/heremaps/oss-review-toolkit/build/1.0.3415#L4375?

Comment (mnonnenmacher) [reply utils/src/main/kotlin/Utils.kt]: Good idea, done.

File: scanner/src/funTest/assets/file-counter-expected-output-for-analyzer-result.yml [hunk 3]
```diff
@@ -196,10 +196,8 @@ scanner:
           errors:
           - timestamp: "1970-01-01T00:00:00Z"
             source: "FileCounter"
-            message: "DownloadException: No source artifact URL provided for 'Gradle:com.here.ort.gradle.example:lib:1.0.0'."
-          - timestamp: "1970-01-01T00:00:00Z"
-            source: "FileCounter"
-            message: "DownloadException: No VCS URL provided for 'Gradle:com.here.ort.gradle.example:lib:1.0.0'."
+            message: "\"DownloadException: No source artifact URL provided for 'Gradle:com.here.ort.gradle.example:lib:1.0.0'.\"\
+              , caused by \"DownloadException: No VCS URL provided for 'Gradle:com.here.ort.gradle.example:lib:1.0.0'.\""
```
Comment (sschuberth) [inline scanner/src/funTest/assets/file-counter-expected-output-for-analyzer-result.yml]: This now needs to be adjusted to "\nCaused by: ", or?

File: utils/src/main/kotlin/Utils.kt [hunk 4]
```diff
@@ -392,6 +392,11 @@ fun Throwable.collectMessages(): List<String> {
     return messages
 }
 
+/**
+ * Recursively collect the messages of this [Throwable] and all its causes and join them to a single [String].
+ */
+fun Throwable.collectMessagesAsString() = collectMessages().joinToString("\nCaused by: ")
```
Comment (fviernau) [inline utils/src/main/kotlin/Utils.kt]: Consider using a bit shorter name, e.g.: getMessagesString() or getMessagesAsString()

File: analyzer/src/main/kotlin/PackageManager.kt [hunk 5]
```diff
@@ -191,6 +191,8 @@ abstract class PackageManager(
                 } catch (e: Exception) {
                     e.showStackTrace()
 
+                    log.error { "Resolving dependencies for '${definitionFile.name}' failed with: ${e.message}" }
```
Comment (fviernau) [inline analyzer/src/main/kotlin/PackageManager.kt]: Not sure if this is out of scope: for the pattern you consistently establish here, would it be possible to reduce the amount of calls from two to one? One option could be to overload log.error as

log.error( message: String, e :  Throwable) {
     e.showstacktrace
     log.error( message )
}

Acknowledge that it''s also not 100% clean to have this showstacktrace call inside a log.* function,
on the positive side it gives more compact client code.

File: scanner/src/funTest/assets/file-counter-expected-output-for-analyzer-result.yml (hunk 3 above)
Comment (mnonnenmacher) [reply scanner/src/funTest/assets/file-counter-expected-output-for-analyzer-result.yml]: Yes, but I expect more tests to fail, so I wanted to wait for the CI results.

File: utils/src/main/kotlin/Utils.kt (hunk 4 above)
Comment (mnonnenmacher) [reply utils/src/main/kotlin/Utils.kt]: I don't like using `get` when it is not a property accessor, that's why I chose `collect`.

File: analyzer/src/main/kotlin/PackageManager.kt (hunk 5 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/PackageManager.kt]: I plan to do something like this in a separate PR, likely as extension function for `Throwable`.

**Pull request merged**